53 platform sensor

patchworquill opened the pull request:
# Change Summary
1. Initialization of I2C peripheral for reading from the sensor. 
2. Initialization of UART peripheral for communicating data to the host computer as serial (baud rate 115200). 
3. Implementation of platform specification for sensor readings. 
4. A demo task for reading pressure and flow from both FS6122 sensors, and sending the results through the UART bus to the host computer to verify the sensor readings. 
5. Initialization of ADC peripheral for reading knob. No implementation of this peripheral or functionality yet (see note 2). 

## Description
This branch implements an attempt at adding the sensor reading (FS6122 - flow, pressure, temp, RH) to the platform. 

The FS6122 sensor is read via i2c bus. `init_i2c` initializes the I2C bus on pins D14 (SDA) + D15 (SCL). 

`pressureFlowSensor.c` adds the necessary functions
- `int get_inspiratory_pressure()`
- `int get_expiratory_pressure()`
- `int get_inspiratory_flow()`
- `int get_expiratory_flow()`

There's another function, `read_FS6122_sensor` that actually reads from the FS6122 sensor. The function currently also uses the _UART_ bus to communicate via serial to the host computer. This needs to be separated, and put into a UART task for communication to the serial monitor. The UART task should then read the FS6122 values from the `MeasurementStatus` struct in `pressureFlowSensor_private.h` and print that through the serial port. 

## Notes
_Note 1: the `sensor_demo.c` task is currently broken. However, the sensor was successfully reading when the code was in all within the `main.c` file (commit ID: `4a96519b5244967a6364d10e96a2c97cb0ba6405`), which implies that the functionality was broken during refactoring._

_Note 2: in slightly messy fashion, I have also added the ADC initialization on this branch, whereas it would be cleaner to have that be its own branch. The ADC peripheral is not being used by this code, but is simply initialized for used._

## Serial Monitoring
On unix-based systems, use the following command to monitor the devices serial port: 

First, find the device ID of the target board:
`ls /dev/tty.*`

Then, create a serial monitor:
`screen /dev.tty.usbmodem<ID> 115200`

sdiemert commented:
@patchworquill 

A general comment about commit style. It is best to use the imperative voice when authoring commit messages. For example, instead of:

```
Adding missing HAL libraries
```

You could write:

```
Add missing HAL libraries
```

I like to think of framing the message in terms of a sentence like *"The result of this commit is to <message>".*

No need to change the messages you have so far... but consider this moving forward. [Here](https://chris.beams.io/posts/git-commit/) is a guideline for commit messages.

sdiemert commented:
@patchworquill 

I think you should leave out the ADC from this PR. It's a great idea to sanity check the sensor values using an analog values. However, I think it makes the PR more complicated. This sensor reading stuff is already complicated enough.

I suggest we make another issue/task for reading the ADC from the sensor.

sdiemert commented on src/platform/pressureFlowSensor.c:
```diff
@@ -0,0 +1,93 @@
+#include <string.h>
+#include <stdint.h>
+
+uint32_t read_FS6122_sensor(uint8_t i2c_addr, uint8_t reg_addr){
```
Suggest using the scheme defined above where `status_t` is returned and the result is written into a pointer. Like: 

```suggestion
status_t read_FS6122_sensor(uint8_t i2c_addr, uint8_t reg_addr, uint32_t * result){
```

sdiemert commented on src/platform/pressureFlowSensor.c:
```diff
@@ -0,0 +1,93 @@
+    ret = HAL_I2C_Master_Transmit(&i2c1_bus, i2c_addr, buf, 1, HAL_MAX_DELAY);
+    if ( ret != HAL_OK ) {
+      strcpy((char*)buf, "Error Tx\r\n");
+      val = 128;
```
Do not set the return value if there was an error. Should remain zero (or some default value).

sdiemert commented on src/platform/pressureFlowSensor.c:
```diff
@@ -0,0 +1,93 @@
+    } 
+    else {
+      // Read 4 bytes from the pressure register
+      ret = HAL_I2C_Master_Receive(&i2c1_bus, i2c_addr, buf, 4, HAL_MAX_DELAY);
```
Use an appropriate delay amount.

patchworquill replied:
What is appropriate in this context? 1s, 100ms, 10ms...? Probably not more than 1s so I'll set that as the max delay for the function call.

patchworquill replied:
set to 1s

LaureMillet replied:
as per the performance requirements each parameter should be less than 1/10 second so the delay should be at most this value. 
In order to define this delay we should have a timing evaluation on how much time is really needed for the sensors, and an evaluation for the task that will use this.

sdiemert commented on src/platform/pressureFlowSensor.c:
```diff
@@ -0,0 +1,93 @@
+      // Read 4 bytes from the pressure register
+      ret = HAL_I2C_Master_Receive(&i2c1_bus, i2c_addr, buf, 4, HAL_MAX_DELAY);
+      if ( ret != HAL_OK ) {
+        strcpy((char*)buf, "Error Rx\r\n");
```
As above, return an error status code if there was a problem interacting with the I2C bus.

patchworquill replied:
This function returns the actual value `uint32_t` when there is a correct read, as this is the general I2C read function. 

I was passing specific values (256, 128) for cases Error_Tx and Error_Rx, and set these two as dummy values that should be changed to something outside the failing range of the sensor, but specific enough that it can trigger a debug / meaningful log entry.

sdiemert commented on src/platform/pressureFlowSensor.c:
```diff
@@ -0,0 +1,93 @@
+      else {
+
+        //Combine the bytes
+        val = ((int16_t)buf[0] << 24) | (buf[1] << 16) | (buf[2] << 8) | (buf[3]) >> 2; // value is 14 bit, so shift by 2
```
`buf` is overloaded here. In one spot it used for as a UART transmission buffer and in other its for tx/rx over i2c. I think this is a bit confusing. We aren't pressed to save memory here, so let's try to make the code easier to understand and use separate buffers.

sdiemert commented on src/platform/pressureFlowSensor.c:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    // Send out buffer (sensor reading or error message)
+    HAL_UART_Transmit(&serial1, buf, strlen((char*)buf), HAL_MAX_DELAY);
```
Remove all UART code from this function. If a user wants to debug, then we should make a separate UART API in the platform that provides all of the required services to print over UART.

sdiemert commented on src/platform/pressureFlowSensor.c:
```diff
@@ -0,0 +1,93 @@
+ *
+ */
+int get_inspiratory_pressure(){
+    return read_FS6122_sensor(INSPIRATORY_ADDR, FS6122_READ_PRESSURE);
```
Isn't there supposed to be data validation in these functions?

sdiemert reviewed (commented):
@patchworquill 

There are a lot of comments here (I tend to do that...). But please don't let that deter you. This is actually a reasonably good first draft! Given how you portrayed it on our call earlier today, I had expected worse ;-).  Consider my excessive commenting a vote of confidence that this PR is headed in the right direction. Also, there are a lot of nitpicking comments.

Some things I like about this: 

1. The general pattern for the platform and application layers is well used.
2. The contribution is reasonably well commented.
3. Code style and naming is good. No random or undecipherable variable names (that I saw at least).

I think I might have found the bug that was preventing your demo task from working. I think you are not properly initializing the i2c structure. The `init_i2c` in `init.c` uses a local (on the stack) variable that is not accessible outside of that function. Then, when you go to use the i2c in `pressureFlowSensor.c` you are using an uninitialized value. See my specific comment in `init.c` for this.

I think some big things that need attention (you probably know this... just listing them for completeness): 

1. The platform APIs for sensor reading should return a `status_t` and write the actual data value into a spot into a pointer passed to the function. I left several comments about this.
2. There is no data validation in the `pressureFlowSensor.c` implementations, do you intend to implement this? The docstrings in the header file seem to indicate you will?
3. Drop the ADC for now. Its out of scope for this, let's implement that later in a separate PR.
4. I know the UART is important for debugging... can this somehow be factored out into it's own API and only pulled into the build in "dev" or "debug" mode?

patchworquill commented:
> @patchworquill
> 
> There are a lot of comments here (I tend to do that...). But please don't let that deter you. This is actually a reasonably good first draft! Given how you portrayed it on our call earlier today, I had expected worse ;-). Consider my excessive commenting a vote of confidence that this PR is headed in the right direction. Also, there are a lot of nitpicking comments.
> 
> Some things I like about this:
> 
> 1. The general pattern for the platform and application layers is well used.
> 2. The contribution is reasonably well commented.
> 3. Code style and naming is good. No random or undecipherable variable names (that I saw at least).
> 
> I think I might have found the bug that was preventing your demo task from working. I think you are not properly initializing the i2c structure. The `init_i2c` in `init.c` uses a local (on the stack) variable that is not accessible outside of that function. Then, when you go to use the i2c in `pressureFlowSensor.c` you are using an uninitialized value. See my specific comment in `init.c` for this.
This sounds like the bug in question. I'll fix this and get it tested. I was having troubles previously with global i2c handler struct definition, but I thought I had fixed it. Hopefully whatever I did for the UART bus can work for the I2C bus initialization. 

> 
> I think some big things that need attention (you probably know this... just listing them for completeness):
> 
> 1. The platform APIs for sensor reading should return a `status_t` and write the actual data value into a spot into a pointer passed to the function. I left several comments about this.

Great note. Will do, was planning on passing the sensor read data into the `MeasurementStatus` struct declared in `pressureFlowSensor_private.h`

> 2. There is no data validation in the `pressureFlowSensor.c` implementations, do you intend to implement this? The docstrings in the header file seem to indicate you will?

Will add the implementation of this tomorrow and (re?)submit the pull request.

> 3. Drop the ADC for now. Its out of scope for this, let's implement that later in a separate PR.
I agree. I had added it originally before considering that the scope of this should be limited to only the sensor. 

> 4. I know the UART is important for debugging... can this somehow be factored out into it's own API and only pulled into the build in "dev" or "debug" mode?
Not exactly sure about how to implement this in terms of the 'dev' / 'debug' mode. Presumably using github's tagging system we can do that, and alter releases accordingly. 
For now, as per some comments I had added, I'll factor the UART stuff into it's own API for use and testing.

sdiemert commented on include/stm32f4xx_hal_conf.h:
```diff
@@ -152,7 +152,7 @@
 #define  DATA_CACHE_ENABLE            1U
 
-#define  USE_HAL_ADC_REGISTER_CALLBACKS         0U /* ADC register callback disabled       */
+#define  USE_HAL_ADC_REGISTER_CALLBACKS         1U /* ADC register callback disabled       */
```
If the ADC is out, should this not be removed?

patchworquill replied:
Yep, didn't see all the extra stuff hiding in this file.

sdiemert commented on include/stm32f4xx_hal_conf.h:
```diff
@@ -36,7 +36,7 @@
 #define HAL_MODULE_ENABLED
 
-  /* #define HAL_ADC_MODULE_ENABLED   */
+  #define HAL_ADC_MODULE_ENABLED
```
Revert to disabled?

sdiemert commented on src/stm32f4xx_hal_msp.c:
```diff
@@ -1,121 +1,252 @@
+    PB8     ------> I2C1_SCL
+    PB9     ------> I2C1_SDA
+    */
+    GPIO_InitStruct.Pin = GPIO_PIN_8|GPIO_PIN_9;
```
I am a bit confused about why this code is here, shouldn't this be in `init.c`?

patchworquill commented on include/platform/init.h:
```diff
@@ -14,6 +14,39 @@
+ * the uart (used for serial monitoring during build and 
+ * debug stages) for the platform has been initialized.
+ */
+void init_uart(void);
```
Any updates on this thought? I can remove it and try testing solely via the debugger (ST Link) for the time being?

sdiemert replied:
I think you could keep the changes locally for testing purposes, and then just omit them from the PR. I see the value of the UART here, but it would be nice to contain the scope as much as possible.

patchworquill commented on include/platform/pressureFlowSensor.h:
```diff
@@ -10,40 +10,54 @@
 #ifndef INC_PLATFORM_PRESSURE_FLOW_SENSOR_H_
 #define INC_PLATFORM_PRESSURE_FLOW_SENSOR_H_
```
Renamed to sensor.h, as this is the only sensor type involved in this model of gVent. Since the sensor also has pressure and temperature, sensor encompasses a bunch of embedded sensors within the single FS6122 sensor unit.

sdiemert commented on src/platform/serial.c:
```diff
@@ -0,0 +1,49 @@
+ */
+status_t print_pressure(uint32_t val){
+    
+<<<<<<< HEAD
```
This shoudl be a syntax error? 

Also, remove this file.

patchworquill replied:
Yes.

sdiemert reviewed (approved):
Approving this so that we can get it into master and proceed with other platform development work. As far as I know, this code has not been fully tested with the real sensor hardware, this will be important for establishing that the sensor code works as expected. However, this can occur at a later time.
